Revamped Notices feature

imath opened the pull request:
The goal of this PR is to transform the code used into the Private Messages component (*an optional component => only available when the Private Messages component is active*) to send all community members a “SiteWide Notice” into a Members component (*a required component => always available*) feature so that:

- The BP Team can use it to inform Administrators about important changes: this new Members Feature is replacing the Admin Notifications workaround we introduced in 11.4.0 to communicate about 12.0.0 important changes.
- Site administrators can notice all community members or all contributors or all other administrators about important informations dealing with community, contribution, or administration matters.
- Integration with the Notifications component is improved as Notices behave similarly to Notifications => they “disappear” once they are dismissed by the user.
## Notices: a Members component feature

While I first thought Notices should be a Core component feature, I quickly changed my mind when working on code changes. The Core component is not setting [needed globals](https://github.com/buddypress/buddypress/blob/master/src/bp-core/classes/class-bp-component.php#L345|L448) such as DB table/table meta and after more thoughts I believe Notices have more to do with Members.

The Members Notices can be deactivated using this filter:

```php
add_filter( 'bp_is_members_notices_active', '__return_false' );
```

Although I’d strongly recommend not to disable it, we’ll always make some adventure users unhappy with it…

A notice is a piece of informations targeting specific member roles:

- `community` All members are targeted by the notice.
- `contributors` Only members able to write posts are targeted by the notice.
- `admins` Only administrators are targeted by the notic.

A notice can have 5 priority levels. 2 levels are not available to administrators:

- `0`  the “Very High” priority level is restricted to the BuddyPress team to inform Administrators about important plugin changes.
- `1` the “High” priority level can be used to move a notice to the top of the notices list.
- `2` the “Regular” priority level is the default one.
- `3` the “Low” priority level can be used to move a notice at the bottom of the notices list.
- `127` is a specific level used to deactivate a notice (just like it was possible within the Private Messages component)

A notice can include a call to action as long as the destination URL has the same domain name than the WordPress site.

Notices are stored into the `wp_bp_notices` table and data about the list of users who dismissed a notice is no longer inside a serialized array of the corresponding user meta but stored as rows into the `wp_bp_notices_meta` table. There’s one row for a user/notice and the name of the Notice meta is `dismissed_by`: Notice ID (`notice_id`  field of the table) was dismissed by (`meta_key` field of the table) User ID (`meta_value`  field of the table).
The BuddyPress DB Schema was updated to migrate data inside the `wp_bp_messages_notices` table if it exists into the new `wp_bp_notices` table.

**Notices management**
Administrators can manage notices from the `Users > Member notices` WP Dashboard menu. From there they can add, edit, deactivate or delete all the notices they created. BuddyPress “Very High” notices cannot be edited or created by the administrators. They can deactivate this kind of notices but they need to wait for all administrators to dismiss it before being able to delete it.

**The BP Notices center**
An orange bell with the number of unread notices/notifications is added to the WP Admin Bar only if there is a notice or a notification (when the Notifications component is active) to display.

Once the orange bell is clicked an HTML Popover is appearing: the first part of it is displaying a “Community notices” section and when the Notifications component is active, it also includes a “Personal notifications” section.

> [!NOTE]
> If users browser doesn’t support the [HTML Popover API](https://caniuse.com/?search=popover), they are redirected to the Notices Member’s front-end profile area when clicking on the orange bell.

> [!NOTE]
> In other words, the Notifications component is using this BP Notices Center to display the logged in member’s notifications. It doesn’t create any new node to the WP Admin Bar anymore and the `src/bp-notifications/bp-notifications-adminbar.php` file is deprecated since 15.0.0.

The “Community notices” section is a “no JS” slider allowing the logged in member to slide up to 5 top priority notices. If there are more than 5 active notices a link is added to the slider’s pagination to view all notices from the Notices Member’s front-end profile area.

A notice can always be dismissed from there thanks to the corresponding “Dismiss” button. If the notice contains a call to action, another button is added for it and clicking on this button will first dismiss the notice before redirecting the user to the call to action URL.

BuddyPress is using this call to action when noticing an upgrade to a major release was successful so that the administrator can directly open the BP Hello/changelog modal to discover what’s new. As a reminder this modal is only automatically displayed after a successful install.

> [!NOTE]  
> BuddyPress Admin Notices are still displayed into the Notices tab of the BuddyPress settings administration screen. From there it’s possible to view dismissed notices as well.

**Notices Member’s front-end profile area**
All notices are by default available into a Notices navigation of the user’s profile. When the Notifications component is active this navigation is moved as a Notifications sub navigation. In this case the Notifications count also takes in account the Notices count. The Notices loop is paginated 5 by 5 and can be filtered to list all unread or dismissed notices. Both Legacy and Nouveau template packs were updated to include notices templates.

## The component’s feature API (new!)

The PR includes a new `BP_Component_Feature` class. Looking back at how we implemented the Site Invitations feature, I believed this part needed to be improved. For Site Invitations we created a component using the `BP_Members_Invitations_Component` object just to be able to have a primary nav into the member’s profile area. The downside of this is it’s creating an active component into the `buddypress()` global. Extending this new `BP_Component_Feature` class which itself extends the `BP_Component` class we avoid the downside and have better control on how to build feature specificities (see `BP_Members_Notices_Feature`). I believe we should go a bit further grouping features files into a feature’s directory. For now I just did the same than we did before: group all functions, template tags etc.., into a single `bp-members-notices.php` file (`bp-{component-name}-{feature-name}.php`).

## The Notices Rest Controller

> [!TIP]
> Until BP REST API v2 is merged into Core, you’ll need to use the following code snippet within a [bp-custom.php](https://github.com/buddypress/buddypress/blob/master/docs/user/advanced/functionalities.md) file to be able to test the JavaScript Powered dismiss action:

```php
function force_notices_feature_rest_api( $feature_class ) {
    add_action( 'bp_rest_api_init', array( $feature_class, 'rest_api_init' ), 10 );
}
add_action( 'bp_members_notices_setup_actions', 'force_notices_feature_rest_api' );
```

I remember having discussions with @renatonascalves about the best way to deprecate BP REST API v1 and class naming for the BP REST API v2. I was confronted to it and after more thoughts, I believe we should name v2 REST API Class controllers like this `BP_{component}_{feature}_REST_Controller` (I probably changed my mind there, sorry!): one of benefit for this is we are following the class autoloader regular map and don’t need to add entries for irregular ones. Here’s the deprecation strategy I used:

- In `BuddyPress->autoload()` keep deprecated endpoint class, eg: `'BP_REST_Sitewide_Notices_Endpoint' => 'messages', // deprecated.`
- Keep & deprecate the old class file, see `src/bp-messages/classes/class-bp-rest-sitewide-notices-endpoint.php` for an example.

The `BP_Members_Notices_REST_Controller`  will greatly benefit from @renatonascalves review & improvements.

Thanks in advance for your tests & reviews!

> [!IMPORTANT]
> Go to `site.url/wp-admin/` first to test the PR, the upgrade routine needs to be accomplished first, you'd get DB errors otherwise.

Trac ticket: https://buddypress.trac.wordpress.org/ticket/9098

---
**This Pull Request is for code review only. Please keep all other discussion in the BuddyPress Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the WordPress Core Handbook for more details.**

imath committed:
Init SiteWide notices migration to the Members component

imath committed:
Add the "Community Notices" profile nav

- Rename "Site Wide Notices" to "Community Notices"
- Make sure this new nav item is shown even if the Extended Profile is disabled.

imath committed:
Add an install routine for Community notices.

imath committed:
Deprecate `BP_Messages_Notice` in favor of `BP_Members_Notice`

imath committed:
Deprecate Messages classes, actions, screens in favor of Members ones.

imath committed:
Make sure the Admin UI to create notice is available

Rearrange the display using 2 cols (like WP Tags or Cats)

imath committed:
Stop using front end to manage Notices

imath committed:
Make sure to display notice bubble even if notifications are inactive

imath committed:
Deprecate `messages_send_notice()` in favor of `bp_members_send_notice()`

- Improve `BP_Members_Notice::save()`
- Use the block API to save richer notice content

imath committed:
Use a popover to display the Member active notice

Deprecate some BP Messages template tags in favor of BP Members ones.

imath committed:
Improve popover display

imath committed:
Move the SiteWide Notice block into the Members component

imath committed:
Improve the Notices center display

imath committed:
Deprecate `bp_messages_dismiss_sitewide_notice()` in favor of `bp_members_dismiss_notice()`

imath committed:
Move Admin Bar submenu generation out of Members component

Avoid notice error removing the blocks PHP included script from the Messages component."

imath committed:
Start adding the Notices metadata API

imath committed:
Adds an upgrade routine for closed notices user meta

imath committed:
Replace 14.0.0 references in favor of 15.0.0 ones

imath committed:
Remaining 14.0.0 references to replace by 15.0.0 ones

imath committed:
Replace "community" wording by "members" one

imath committed:
Restore 2 14.0 deprecated functions

imath committed:
Improve code indentation

imath committed:
Add the 15.0 updater to the BP upgrading process

imath committed:
Use a meta query to get dismissed notices

imath committed:
Remove the `BP_Members_Admin->get_heading_action()` method

Improve the table upgrade to use `priority` instead of `is_active`

imath committed:
Introduce the `bp_members_get_user_higher_priority_notice()` function

imath committed:
Fix some merge fusions mistakes

If the community profile is disabled, no need to create the profile navigation and load CSS/JS.

Fix another merge fusion mistake

No need to add the check to see if community profil is enabled.

imath committed:
Remove user's notice admin screen

I change my mind, in WP Admin, the Admin bar is available.

imath committed:
Move the update task for bp notice table

imath committed:
Improve Popover display on front-end

imath committed:
Introduce a `BP_Component_Feature` class to organize component features

imath committed:
Deprecate notice tags in BP Nouveau

imath committed:
Adds a Notices submenu to My Account Admin bar menu

imath committed:
Add a DB Table column to store notice types

imath committed:
Use "target" instead of "type"

imath committed:
Add a field to get the action button text

Remove messages deprecated filters & Add a form field to define priority.

imath committed:
Use priority 255 to deactivate a notice
Add a column to display priority into the Notices Admin screen

imath committed:
Add the `bp_notices_delete_meta` function.
Use 127 as the deactivated priority.
Init Edit Notice Admin screen.

imath committed:
Start using one form for both screens

imath committed:
Make sure the form is filled with notice values to update

imath committed:
MIgrate Admin notifications

imath committed:
Dismiss Admin notices during migration

Introduce a function to get all administrator IDs.

imath committed:
Output Admin Notices in BP Admin Screen

imath committed:
Manage Notices count queries

imath committed:
Make the Notifications component use the Notices center

imath committed:
Avoid a fatal error when trying to count `false`

`bp_notifications_get_notifications_for_user()` is returning `false` when the user has no notices.

imath committed:
Move the notices screen into the Notifications component if active

imath committed:
Start building Notices templates

Adds notices templates to BP Nouveau.

imath committed:
Use `bp_members_get_notices()` to loop on front end

imath committed:
Handle dismissing a notice

Adds an action to handle it that can be used from Admin & Front End.
Deprecate functions About Admin Notifications
Remove JS to dismiss admin notifications.

imath committed:
Add a notice for fresh & upgraded installs

imath committed:
Make it possible to redirect to cta URL once a notice is dismissed

imath committed:
Allow BP Admin notice to be deleted (once every admin read it)

imath committed:
List up to 5 notices into the notices center

imath committed:
Display notices like a slider inside the notices center

imath committed:
Remove Target Dashicons in favor or a title prefix

imath committed:
Add pagination to Notices Center

- Add single notice views on front & back end
- Introduce a template tag to get the notice excerpt to makr sure content doesn't exceed 300 characters.

imath committed:
Fix failing unit tests

imath committed:
PHP Unit tests avoid deprecation notices

imath committed:
Make BP Legacy supports Members Notices

imath committed:
Adapts Sitewide Notice block to Members Notices

imath committed:
Add a fallback for browsers not supporting the Popover API

imath committed:
Deprecate all remaining Messages functions relative to Notices

imath committed:
Nothing personal!

imath committed:
Adds new controller for Notices REST API requests

imath committed:
Introduce a new JS Notices controller to use REST API Requests

Update the Notices Center script to use it to dismiss a notice.

imath committed:
Catch REST API dismiss request errors

imath committed:
Update the Notices Center top priority pagination on dismiss, using JS

imath committed:
Admin notices: add paginated results

imath committed:
Remove extra unclosed form in Admin Notices screen.

imath committed:
Include Notices count into BP Nav an WP Admin Bar items

imath committed:
Improve notice display on front end

Add pagination and filter

imath committed:
Improve Notices center display on mobiles

imath committed:
Improve Notices display in BP Legacy template pack

imath committed:
Remove scroll bar in slider

imath committed:
Start working on Notices cache

imath committed:
Cache dismissed notices and notices count

imath committed:
Add cache for user's top priority notices

imath committed:
Use the user's notification link when no notices are available

imath committed:
Reorganize functions and rename some (rendering ones)

imath committed:
Fix escaping errors

imath committed:
Set collection params for Notices endpoint

imath committed:
Fix errors revealed using `grunt commit`

imath committed:
Improve the JS to dismiss notice from the SiteWide notice block

imath committed:
Last touches before pulling a request

Remove the action about REST API (the description of the PR will provide a hook to test it)
Improve Notices Center front end display styling buttons like on back-end.

imath committed:
Avoid errors when the Member Notices feature is disabled.

It shouldn't! But some of our users are adventurers 😱

imath committed:
Last commit before PR (really this time!)

- More `bp_is_active()` checks
- update db raw version
- remove a cache group not used anymore

renatonascalves commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,758 @@
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
```
```suggestion
defined( 'ABSPATH' ) || exit;
```

renatonascalves commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,758 @@
+	 *     @type boolean      $count_total_only `true` to only get the total number of notices, `false` otherwise.
+	 *                                          Defaults to `false`.
+	 * }
+	 * @return array|integer|null List of notices, total number of notices for a count only query or null if nothing was found.
```
Most of our `::get`, returns the objects (or ids only) and the total number of matches objects at the same time. I think maybe to avoid a second query. Not all classes support this though.

imath replied:
Most of these `::get` are performing 2 queries: one to list objects, the other to get total, I believe using it this way is avoiding making 2 queries when we only need one 😄

renatonascalves commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,758 @@
+				);
+
+				if ( $is_user_notices ) {
+					wp_cache_set( $args['user_id'], $result, 'bp_member_top_priority_notices' );
```
As part of https://buddypress.trac.wordpress.org/ticket/8552, we should add support for `cache_results`. I can do it after this is merged too.

imath replied:
I agree, let's do it once it's merged 👍

renatonascalves commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,758 @@
+	 *
+	 * @since 15.0.0
+	 *
+	 * @param array $args {
```
Is it possible to `orderby` by something? Priority maybe? Target?

renatonascalves replied:
Also sort.

imath replied:
I prefer to keep the order/orderby readonly for now. Notices are always ordered by priority and date.

```php
ORDER BY priority ASC, date_sent DESC
```

renatonascalves commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,758 @@
+	 *     An array of arguments.
+	 *
+	 *     @type integer      $user_id          The user ID to get notices for. Defaults to `0`.
+	 *     @type integer      $pag_num          Number of notices per page. Defaults to `20`.
+	 *     @type integer      $pag_page         The page number to retrieve.  Defaults to `1`.
```
Can get change this to a more familiar syntax? Like `$per_page` and `$page`.

imath replied:
I agree we should probably move to this syntax, but then we should do it for all BuddyPress objects. I think it's best to deal with it globally from another ticket.

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-admin.php:
```diff
@@ -0,0 +1,405 @@
+/**
+ * BuddyPress members component Site-wide Notices admin screen.
+ *
+ * @package buddypress\bp-members\classes\class-bp-members-notices-admin
+ * @since 15.0.0
```
```suggestion
 * @package BuddyPress
 * @subpackage Members
 * @since 15.0.0
```

renatonascalves commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,758 @@
+ *
+ * @since 15.0.0
+ */
+class BP_Members_Notice {
```
I think this classs needs `#[AllowDynamicProperties]` too.

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-admin.php:
```diff
@@ -0,0 +1,405 @@
+
+			check_admin_referer( 'save-notice', 'ns-nonce' );
+
+			$notice = bp_parse_args(
+				$_POST['bp_notice'],
+				array(
+					'id'       => 0,
+					'title'    => '',
+					'content'  => '',
+					'target'   => '',
+					'priority' => 2,
+					'url'     => '',
+					'text'     => '',
+				)
+			);
```
While testing this, and looking at the code, it seems title and content should be required. But there is no _visual_ indication of that. Try to send a notice without any content and you get a successful message.

![CleanShot 2024-09-02 at 13 14 08@2x](https://github.com/user-attachments/assets/99941f88-c6a3-4828-b7bb-84eeab76e825)

imath replied:
Good catch! I'll improve this.

imath replied:
<img width="1245" alt="user-feedback" src="https://github.com/user-attachments/assets/6205e338-a407-483f-aa97-8eb7a2f44d4c">

When title and/or content are not provided it now displays a more meaningful error + highlights the required fields like WordPress does for categories/tags.

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-admin.php:
```diff
@@ -0,0 +1,405 @@
+				</div><!-- /col-left -->
+				<div id="col-right">
+					<div class="col-wrap">
+						<?php $this->list_table->display(); ?>
```
I think one possible feature request we might get is to show the total number of users that read/dismissed the notice.

![CleanShot 2024-09-02 at 13 16 09@2x](https://github.com/user-attachments/assets/3d8f4889-f797-47af-be07-0d835a4779ce)

imath replied:
I agree, I thought about it myself 👍 . I'll see what I can do about it once most of the PR has been committed 😉

renatonascalves reviewed (changes requested):
✅  I'm done reviewing it! =D

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+			return $fields_update;
+		}
+
+		$retval   = $this->prepare_item_for_response( $notice, $request );
```
This method already sets `rest_ensure_response`. 🤔

renatonascalves replied:
Do you know why we set it again in line 349?

imath replied:
I have no idea to be honest, I haven't touched this part from the BP REST plugin version 🤔

imath replied:
I checked all other BP REST API endpoints as well as the `WP_REST_Posts_Controller`: using the `rest_ensure_response()` twice (inside the `prepare_item_for_response()` & into the request callback) is also practiced. I don't know why, but it looks to be the way to do it... So let's carry on using it twice! As this [function](https://developer.wordpress.org/reference/functions/rest_ensure_response/) is only making sure a `WP_REST_Response` object is returned, it's hurting noone: twice is always better than never!

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function update_item( $request ) {
+		// Setting context.
+		$request->set_param( 'context', 'edit' );
```
I think this should be passed by the consumer of the API. We have been removing this hardcoded values in other REST API endpoints.

imath replied:
I haven't changed how it currently is in https://github.com/buddypress/BP-REST/blob/master/includes/bp-messages/classes/class-bp-rest-sitewide-notices-endpoint.php#L305

Do you remember why you chose to force the context in the first place?

renatonascalves replied:
I **incorrectly** assumed that any `WP_REST_Server::EDITABLE` request use the `edit` context, which is not always the case. It's up to the API consumer to decide which data to show.

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+			)
+		);
+
+		if ( is_user_logged_in() ) {
```
I _think_ it is technically possible for _any_ logged in user to dismiss someone else's notices. As long as they have access to the notice ID. Is this expected?

imath replied:
Notices were targeting every member of the community in the previous implementation, I agree I now need to be me restrictive here to check admin notices can only be viewed by admins and contributor notices can only be viewed by contributors.

imath replied:
I've updated this part this way:
```php
if ( is_user_logged_in() ) {
		// Users can only get notices targeting them.
		if ( 'view' === $context && bp_loggedin_user_id() === (int) $request->get_param( 'user_id' ) ) {
```

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+	 */
+	public function delete_item( $request ) {
+		// Setting context.
+		$request->set_param( 'context', 'edit' );
```
See my feedback about setting this manually.

imath replied:
Yes I have the same question: any idea why you set it this way in the first place?
See https://github.com/buddypress/BP-REST/blob/master/includes/bp-messages/classes/class-bp-rest-sitewide-notices-endpoint.php#L541

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+	 * @return array
+	 */
+	public function get_collection_params() {
+		$params                       = parent::get_collection_params();
```
This is adding some fields which I'm not sure are supported currently. Like `search`.

imath replied:
Good catch, I'll clean up these fields.

imath replied:
Only the `search` field needed to be removed ✅

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+				'enum' => array( 'community', 'contributors', 'admins' ),
+				'type' => 'string',
+			),
+			'sanitize_callback' => 'wp_parse_list',
```
```suggestion
			'sanitize_callback' => 'bp_rest_sanitize_string_list',
```

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+			'description'       => __( 'Limit result set to items having a specific priority.', 'buddypress' ),
+			'type'              => 'integer',
+			'items'             => array(
+				'enum' => array( 1, 2, 3 ),
```
So it is not possible to fetch other types of items as an admin?

imath replied:
Absolutely 😄

renatonascalves commented on src/bp-members/classes/class-bp-members-notices-rest-controller.php:
```diff
@@ -0,0 +1,1090 @@
+		$prepared_item = bp_members_get_notice( $notice_id );
+
+		// Notice ID.
+		if ( ! empty( $schema['properties']['id'] ) && ! empty( $prepared_item->id ) ) {
+			$prepared_item->id = $prepared_item->id;
+		}
```
Oddly, if the `id` property is empty, we are passing it along anyway. So I'm not sure how useful this conditional actually is.

imath replied:
I've edited it to remove the ID if the schema doesn't have the ID property, I guess it's to prepare a create request.

renatonascalves commented on src/bp-messages/classes/class-bp-rest-sitewide-notices-endpoint.php:
```diff
@@ -0,0 +1,38 @@
+
+defined( 'ABSPATH' ) || exit;
+
+_deprecated_file(
+	basename( __FILE__ ),
+	'15.0.0',
+	'bp-members/classes/class-bp-members-notices-rest-controller.php',
+	esc_html__( 'The v1 of the BP REST API is deprecated, please use v2.', 'buddypress' )
+);
```
I'm a but confused about having this class here. If the BP REST plugin is active, I'd expect we would get a PHP error since we would be loading the same class twice.

imath replied:
We had a discussion about it, the important part is to remove the deprecated class from the class autoloader irregular map.

imath replied:
I've removed the deprecated class from the the autoloader.

renatonascalves commented on src/js/blocks/bp-members/webpack.config.js:
```diff
@@ -10,12 +10,22 @@ module.exports = {
     ...defaultConfig,
 	...{
         entry: {
+			'notices-center/controller': {
```
It seems there is a spacing issue here.

imath replied:
I've checked, it doesn't look it does 🤔

renatonascalves commented:
> I remember having discussions with @renatonascalves about the best way to deprecate BP REST API v1 and class naming for the BP REST API v2. I was confronted to it and after more thoughts, I believe we should name v2 REST API Class controllers like this BP_{component}_{feature}_REST_Controller (I probably changed my mind there, sorry!): one of benefit for this is we are following the class autoloader regular map and don’t need to add entries for irregular ones. Here’s the deprecation strategy I used:

Changing the format of the files for the V2 endpoints is no problem at all. I think the main issue before was renaming the files from V1, but since we will bundle it in the bp rest plugin, keeping their names, again, it's no problem at all.

imath commented:
Hi @renatonascalves thanks a lot for your review 😍 . I'll look at it in details asap!

imath committed:
Apply a first round of @renatonascalves recommendations

imath commented on src/bp-members/bp-members-cache.php:
```diff
@@ -125,3 +125,22 @@ function bp_members_reset_signup_cache_incrementor() {
+	wp_cache_delete( $user_id, 'bp_member_notices_count' );
+	wp_cache_delete( $user_id, 'bp_member_top_priority_notices' );
+}
+add_action( 'bp_members_notice_after_save', 'bp_members_clear_user_notices_cache', 10, 0 );
+add_action( 'bp_members_notice_after_delete', 'bp_members_clear_user_notices_cache', 10, 0 );
```
Actually no, because these to hooks are not returning the user ID as their first argument, reason why I used `0`.

imath commented on src/bp-members/bp-members-notices.php:
```diff
@@ -0,0 +1,1510 @@
+		$raw_content = $parsed_content['innerHTML'];
+
+		if ( $raw ) {
+			return $raw_content;
```
Yes it's intentional 😄

imath commented on src/bp-members/bp-members-notices.php:
```diff
@@ -0,0 +1,1510 @@
+function bp_notice_target( $notice = null ) {
+	$target_key = bp_get_notice_target( $notice );
+
+	$targets = array(
```
Good point, we can always improve it once most of the code is merged into Core 😅

imath commented on src/bp-core/classes/class-bp-admin.php:
```diff
@@ -1687,28 +1653,111 @@ public function display_addons_table() {
+				}
+
+				printf(
+					'<div id="message" class="%1$s notice is-dismissible"><p>%2$s</p></div>',
```
We are using the WordPress `id` to inherit admin styles, that's why we don't add a `bp-` prefix. I've checked in other Admin files.

imath committed:
Apply a second round of @renatonascalves recommendations

imath commented on src/bp-core/admin/bp-core-admin-schema.php:
```diff
@@ -632,3 +626,61 @@ function bp_core_install_nonmember_opt_outs() {
+ *
+ * @since 15.0.0
+ */
+function bp_core_install_members_notices() {
```
It would be useful only if the `wp_bp_messages_notices` table still exists, the upgrade routine is renaming it to `wp_bp_notices`. I'm unsure it's worthing the effort, but let's get team's opinion about it!

imath committed:
Improve Notices REST Controller and Administration

- Remove the deprecated `BP_REST_Sitewide_Notices_Endpoint` from autoloaded classes.
- Improve user feedback to inform title and content are required fields when creating a notice from the corresponding Admin screen.
- Make sure users can only get notices targeting them when using the REST API.

Props @renatonascalves

imath commented:
@renatonascalves I believe I've improved most important parts according to your review.

About the 2 remaining REST API points: duplicate usage of `rest_ensure_response` and forcing the `edit` context for create/update/delete requests, I'll do what you confirm is best. I've copied these from the current v1 REST API endpoint.

About the other improvements + adding more unit tests, I'd suggest to work on it once the current code is merged 😅

imath committed:
Leave the API consumer to decide about the request context

imath committed:
Update PR according to latest Trunk.

Removes the Messages Notices REST Controller as goal of this PR is to move Notices into the Members component.

imath committed:
Transfer Messages Notices Unit Tests to Members Notices ones

imath commented:
Just updated the PR, it's now safe to test it again!

You don't need the following tip anymore as the v2 of the BP REST API has been merged into Trunk.

```php
function force_notices_feature_rest_api( $feature_class ) {
    add_action( 'bp_rest_api_init', array( $feature_class, 'rest_api_init' ), 10 );
}
add_action( 'bp_members_notices_setup_actions', 'force_notices_feature_rest_api' );
```
If you added it ^^ to test before, please remove it!

imath committed:
Introduce the `bp_members_admin_bar_use_notifications_menu` filter

Using `add_filter( 'bp_members_admin_bar_use_notifications_menu', '__return_true' );` Theme developers not ready for the Notices center can keep the "Notifications" way of displaying notifications.

imath commented:
To avoid the theme issue @emaralive shared during latest dev-chats, I've just introduced a filter to keep the good old way to display notifications on the front end of the site.

Using:
```php
add_filter( 'bp_members_admin_bar_use_notifications_menu', '__return_true' );
```

Here's how it looks when you only have notifications
<img width="647" alt="Capture d’écran 2024-10-12 à 11 55 11" src="https://github.com/user-attachments/assets/5132b1da-fa10-49e5-b450-8fdd0bcef4aa">

Here's how it looks if you have a least one notice
<img width="582" alt="Capture d’écran 2024-10-12 à 11 55 40" src="https://github.com/user-attachments/assets/bfe2c1c3-2d20-4d46-b453-aff46a535b67">

imath committed:
Update package.json to remove Messages blocks task

imath committed:
Fix php notice, i18n warnings as well as WPCS error

dcavins commented on src/bp-members/bp-members-adminbar.php:
```diff
@@ -165,106 +164,182 @@ function bp_members_admin_bar_user_admin_menu() {
+					'id'      => 'notices',
+					'content' => sprintf(
+						/* translators: %s: total notices count */
+						_n( 'Warning: you have %s unread notice', 'Warning: you have %s unread notices', $notices_count, 'buddypress' ), number_format_i18n( $notices_count )
```
"Warning" is a little too strong of a word. How about "Important: You have %s unread notice"? (The American English style is to often capitalize the first word after a colon if the phrase is a sentence.)

dcavins commented on src/bp-core/deprecated/15.0.php:
```diff
@@ -0,0 +1,391 @@
+ */
+function bp_notices_clear_cache() {
+	_deprecated_function( __FUNCTION__, '15.0.0' );
+}
```
Wow! So many deprecations. Are we reasonably confident that themes haven't built their own notices interface using these functions?

dcavins commented on src/bp-members/bp-members-notices.php:
```diff
@@ -0,0 +1,1515 @@
+
+	remove_filter( 'query', 'bp_filter_metaid_column_name' );
+
+
```
Extra line return here.

dcavins commented on src/bp-members/classes/class-bp-members-notice.php:
```diff
@@ -0,0 +1,760 @@
+				)
+			);
+
+			// Cache user's first active notice.
```
Cache user's dismissed notices.

dcavins commented:
This is an excellent change, and one that I appreciated more after seeing how it improves the focus of the existing messages and notifications code. Wow, what a lot of work! Thanks for doing this! It should make all of these somewhat related pieces easier to maintain and improve. :)

emaralive commented:
> To avoid the theme issue @emaralive shared during latest dev-chats, I've just introduced a filter to keep the good old way to display notifications on the front end of the site.
> 
> Using:
> 
> ```
> add_filter( 'bp_members_admin_bar_use_notifications_menu', '__return_true' );
> ```
> 
> Here's how it looks when you only have notifications <img alt="Capture d’écran 2024-10-12 à 11 55 11" width="647" src="https://private-user-images.githubusercontent.com/1834524/375948795-5132b1da-fa10-49e5-b450-8fdd0bcef4aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwODA0MDYsIm5iZiI6MTczMTA4MDEwNiwicGF0aCI6Ii8xODM0NTI0LzM3NTk0ODc5NS01MTMyYjFkYS1mYTEwLTQ5ZTUtYjQ1MC04ZmRkMGJjZWY0YWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDhUMTUzNTA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODMwMWY5ODQ0Y2Y4NDIwZDUyYTA1NjczMmQ1MGU3YWFjYTZkYTk5ZDg5ZjExZGNmZGIwMDI4ZTI4MDgyYTljNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.SrX94m_WN6spHOUCJDgN40klkLET2K194bePzD-Hs2M">
> 
> Here's how it looks if you have a least one notice <img alt="Capture d’écran 2024-10-12 à 11 55 40" width="582" src="https://private-user-images.githubusercontent.com/1834524/375948807-bfe2c1c3-2d20-4d46-b453-aff46a535b67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwODA0MDYsIm5iZiI6MTczMTA4MDEwNiwicGF0aCI6Ii8xODM0NTI0LzM3NTk0ODgwNy1iZmUyYzFjMy0yZDIwLTRkNDYtYjQ1My1hZmY0NmE1MzViNjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDhUMTUzNTA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjUzZTMxNzVjMWI3MGZhMWQzNWI4M2U2ZmNlNzAyNDA1YzZhMjNjOTc4NDc5YTRhODVkZjU1NmE5NDgzMTRhYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.JQlG3djVg004Ll3fPXIIymxHUKC2AyN516k4FhOGv5M">

The filter attempts to apply a _band-aid_ ( :arrow_left:  for lack of a better term) as a solution for the issue that can be seen in the following image (same image shown in slack when I raised the issue:

![Git-WP-Built-by-Photographers-for-Photographers-10-02-2024_12_52_PM](https://github.com/user-attachments/assets/32a1abc1-95be-4e6d-bd9c-4d73dff8a916)

However, the indicated filter assumes the toolbar is available on the front-end which is not the case for all sites, as can be demonstrably seen. Applying the filter, in the case of "**no toolbar**" looks like the following (no indication or access to notifications because the toolbar is not present):

![bp_members_admin_bar_use_notifications_menu-true](https://github.com/user-attachments/assets/1c4341b5-2892-4624-a671-fd9cad7a8812)

If a user has access to their profile area within wp-admin, they can turn off the toolbar or the site  admin can turn off the toolbar universally with:

```
add_filter( 'show_admin_bar', '__return_false' );
 ```

What's also interesting is that when the "**show_admin_bar**" filter returns a **false** or when the "**Show Toolbar when viewing site**" checkbox is disabled within a user's profile (wp-admin) there are no **notices** on the front-end, regardless of whether the "**bp_members_admin_bar_use_notifications_menu**" filter returns a **true**. This infers that a dependency exists of some sorts and TBH, I've not examined the code, this is just testing various scenarios with the **TwentyEleven** theme.  Another bit of information, the reason notices can be seen when there appears to be "**no toolbar**" is due to CSS is being used to hide the **toolbar**.

A final note for this comment is that there is an additional issue when the "**bp_is_members_notices_active**" returns a **false**, as can be seen in this next image:

![bp_is_members_notices_active-false](https://github.com/user-attachments/assets/f514382e-a065-423f-a001-304b2abf3d6c)

More testing is required with other themes and IMHO, this isn't ready to be released with 15.0.
